baseplate/experiments/__init__.py

```diff
@@ -0,0 +1,171 @@
+            to None to disabled bucketing events entirely.
+        """
+        self._filewatcher = FileWatcher(path, json.load)
+        self._event_queue = event_queue
+
+    def make_object_for_context(self, name, server_span):
+        return Experiments(self._filewatcher, self._event_queue)
+
+
+class Experiments(object):
+    """ Access to experiments with automatic refresh when changed. Handles
+    logging bucketing events to the event pipeline.
+    """
+
+    def __init__(self, config_watcher, event_queue=None):
+        self._config_watcher = config_watcher
+        self._event_queue = event_queue
+        self._already_bucketed = set()
+
+    def _get_config(self, name):
+        try:
+            config_data = self._config_watcher.get_data()
+            return config_data[name]
+        except WatchedFileNotAvailableError:
+            logger.warning("Experiment config file not found")
+            return
+        except KeyError:
+            logger.warning(
+                "Experiment <%r> not found in experiment config",
+                name,
+            )
+            return
+        except TypeError:
+            logger.warning("Could not load experiment config: %r", name)
+            return
+
+    def variant(self, name, bucketing_event_override=None,
+                extra_event_params=None, **kwargs):
+        """ Which variant, if any, is active for the experiment specified by
+        "name".  If a variant is active, a bucketing event will be logged to
+        the event pipeline unless any one of the following conditions are met:
+
+        1. bucketing_event_override is set to False.
+        2. The experiment specified by "name" explicitly disables bucketing
+           events.
+        3. We have already logged a bucketing event for the value specified by
+           experiment.event_cache_key(**kwargs) within the current session.
+
+        :param str name: Name of the experiment you want to run.
+        :param bool bucketing_event_override: Optional value, defaults to None.
+            If set to True, will always log bucketing events unless the
+            experiment explicitly disables them.  If set to False, will never
+            send a bucketing event.  If set to None, no override will be
+            applied.
+        :param dict extra_event_params: Optional value.  Any extra values you
+            want to add to the bucketing event.
+        :param kwargs:  Arguments that will be passed to experiment.variant to
+            determine bucketing, targeting, and overrides.
+
+        :rtype: :py:class:`str`
+        :return: Variant name if a variant is active, None otherwise.
+        """
+        config = self._get_config(name)
+        if not config:
+            return None
+
+        experiment = parse_experiment(config)
+        variant = experiment.variant(**kwargs)
+
+        do_log = True
+
+        if variant is None:
+            do_log = False
```

Comment:
Not if we want to do what r2 currently does, if you set `bucketing_event_override=True` it will still send one even if `variant is None`.

```python
        should_send_bucketing_event = False
        if variant is None or bucketing_event_override is False:
            # Never try sending if there's no variant or override is False
            should_send_bucketing_event = False
        elif bucketing_event_override:
            # Always try sending if override is truthy
            should_send_bucketing_event = True
```

I personally have no attachment to that behavior and am fine with removing it.